internal/aws/cwlogs/pusher_cache.go

```diff
@@ -0,0 +1,95 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cwlogs // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/cwlogs"
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type PusherCache interface {
+	// GetPusher will check to see if there already exists a pusher configured to write to
+	// the given stream and group, and that pusher will be returned.
+	// If no pusher exists for a given stream and group, a new one is created, stored and returned
+	GetPusher(logGroup, logStream string, client Client, retries int) (pusher Pusher)
+
+	// Shutdown will force each stored pusher to flush any data currently buffered
+	// and then clear the existing cache.
+	Shutdown(ctx context.Context) error
+
+	// ListPushers will return a list of all pushers in the cache.
+	ListPushers() []Pusher
+
+	// Flush will cause all pushers to flush to the AWS API
+	Flush() error
+}
+
+type DefaultPusherCache struct {
```

Comment:
I would suggest not exporting this time since the zero value isn't safe for use.

Reply:
huh? I am not sure what you mean here?  Everything that I export is required by other code.

Reply:
Do you mean th cache? it has to be exported, its used in other packages.

Reply:
Sorry, I meant to say `type` instead of `time`.

Reply:
The type it shouldn't should not need to be exported since it implements `PusherCache` interface.

Reply:
I use the interface in unit tests for easy mock, hence public

Reply:
Sorry for not clearly explaining myself here but `DefaultPusherCache` should not be exported since it's type zero value is unsafe, meaning:

```golang
pc := &DefaultPusherCache{}
pusher := pc.GetPusher(...) // Panic
```

The reason being is that now internal values have been set and are using their zero values.

Reply:
The `DefaultPusherCache` type does not need to be exported, and shouldn't be for the reasons that @MovieStoreGuy points out.  Instead, create an initializer for it that is the only way for an external package to get an instance of it:

```go
type defaultPusherCache struct {...}

func NewPusherCache(...) PusherCache {
  return &defaultPusherCache{logger: logger, ...}
}
```

Something like this will ensure that the unsafe zero value can't be used.